src/value.c

```diff
@@ -371,58 +371,59 @@ SIValue SIValue_Divide(const SIValue a, const SIValue b) {
 	return SI_DoubleVal(SI_GET_NUMERIC(a) / (double)SI_GET_NUMERIC(b));
 }
 
-int SIArray_Compare(SIValue arrayA, SIValue arrayB) {
+int SIArray_Compare(SIValue arrayA, SIValue arrayB, int *disjointOrNull) {
 	uint arrayALen = SIArray_Length(arrayA);
 	uint arrayBLen = SIArray_Length(arrayB);
-	// check empty list
+	// Check empty list.
 	if(arrayALen == 0 && arrayBLen == 0) return 0;
 	int lenDiff = arrayALen - arrayBLen;
-	// check for the common range of indices
-	uint minLengh = arrayALen <= arrayBLen ? arrayALen : arrayBLen;
+	// Check for the common range of indices.
+	uint minLength = arrayALen <= arrayBLen ? arrayALen : arrayBLen;
 
-	uint allDisjoint = 0;   // counter for the amount of disjoint comparisons
-	uint nullCompare = 0;   // counter for the amount of null comparison
-	// notEqual holds the first false (result != 0) comparison result between two values from the same type, which are not equal
+	uint notEqualCounter = 0;   // Counter for the amount of false (compare(a,b) !=0) comparisons.
+	uint nullCounter = 0;       // Counter for the amount of null comparison.
+	// notEqual holds the first false (result != 0) comparison result between two values from the same type, which are not equal.
 	int notEqual = 0;
 
 	// Go over the common range for both arrays.
-	for(uint i = 0; i < minLengh; i++) {
+	for(uint i = 0; i < minLength; i++) {
 		SIValue aValue = SIArray_Get(arrayA, i);
 		SIValue bValue = SIArray_Get(arrayB, i);
-		int compareResult = SIValue_Compare(aValue, bValue);
-		switch(compareResult) {
-		case 0:
-			// we have a matching value
-			break;
-		case COMPARED_NULL:
-			// there was a null comparison
-			nullCompare++;  // update null comparison counter
-			allDisjoint++;  // null comparison is also a disjoint comparison
-			break;
-		case DISJOINT:
-			allDisjoint++;  // there was a disjoint comparison
-			if(notEqual == 0) notEqual =
-					compareResult; // if there wasn't any false comparison, update the false comparison to disjoint value
-			break;
-		default:
-			// if comparison is not 0 (a != b), set the first value
+		// Current comparison special cases indication variable.
+		int currentDisjointOrNull;
```

Comment:
Init to `0` do not trust `SIValue_Compare` to init this value.

Reply:
fixed